### ISSUE TITLE
Job never ends cause salt-api returns 500

### ISSUE THREAD

**Issue by albertmfb** [vmware-archive/salt-api#168]
Good morning,
I have problems using the salt-api in one job executed via rundeck with saltstack.
The commans of the job are executed succesfully but job never ends.

product versions:
rundeck-2.6.1-1.12.GA.noarch
salt-master-2014.7.0-3.el6.noarch
salt-api-2014.7.0-3.el6.noarch

The salt-API works fine in all executions except this.
The configuration in  /etc/salt/master for salt-api through cherrypy:

rest_cherrypy:
  port: 8000
  host: 192.168.35.13
  ssl_crt: ....cert.pem
  ssl_key: .....key.pem

In rundeck logs via web I see:

00:15:00    lpgst301.igrupobbva 1. Remote Salt Execution    [workflow] beginExecuteNodeStep(lpgst301.igrupobbva): NodeDispatch: StepExecutionItem{type='NodeDispatch', keepgoingOnSuccess=false, hasFailureHandler=false}
00:15:00            Using salt-api version: [org.rundeck.plugin.salt.version.SaltApiCapability@682ddad6]
00:15:00            Authenticating with salt-api endpoint: [https://xxx.xxx.xxx.xxx:8000/login]
00:15:00            Submitting job with arguments [[fun=cmd.run_all, tgt=lpgst301, arg=runas=nagios, arg=timeout 36000 bash /MONITORIZACION/uti/launcherMaximosMinimos.sh]]
00:15:00            Submitting job with salt-api endpoint: [https://xxx.xxx.xxx.xxx:8000/minions]
00:15:00            Received response for job submission = HTTP/1.1 202 Accepted [Content-Length: 143, Access-Control-Expose-Headers: GET, POST, Access-Control-Allow-Credentials: true, Vary: Accept-Encoding, Server: CherryPy/3.2.2, Allow: GET, HEAD, POST, Cache-Control: private, Date: Tue, 24 Nov 2015 23:15:00 GMT, Access-Control-Allow-Origin: *, Content-Type: application/json, Set-Cookie: session_id=4f6b5526848a09956b88c8d305ec715ed89d0a6c; expires=Wed, 25 Nov 2015 09:15:00 GMT; Path=/]
00:15:00            Received jid [20151125001500479794] for submitted job
00:15:00            Polling for job status with salt-api endpoint: [https://xxx.xxx.xxx.xxx:8000/jobs/20151125001500479794]
02:04:15            Encountered recoverable status code: HTTP/1.1 500 Internal Server Error
02:09:16            Encountered recoverable status code: HTTP/1.1 500 Internal Server Error
02:14:21            Encountered recoverable status code: HTTP/1.1 500 Internal Server Error
02:19:36            Encountered recoverable status code: HTTP/1.1 500 Internal Server Error
02:24:41            Encountered recoverable status code: HTTP/1.1 500 Internal Server Error
02:30:23            Encountered recoverable status code: HTTP/1.1 500 Internal Server Error

In salt logs the only error showed is:
2015-11-25 00:15:52,508 [cherrypy.error                           ][ERROR   ] [25/Nov/2015:00:15:52]  Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/cherrypy/_cpwsgi.py", line 169, in trap
    return func(_args, *_kwargs)
  File "/usr/lib/python2.6/site-packages/cherrypy/_cpwsgi.py", line 96, in __call__
    return self.nextapp(environ, start_response)
  File "/usr/lib/python2.6/site-packages/cherrypy/_cpwsgi.py", line 379, in tail
    return self.response_class(environ, start_response, self.cpapp)
  File "/usr/lib/python2.6/site-packages/cherrypy/_cpwsgi.py", line 222, in __init__
    self.run()
  File "/usr/lib/python2.6/site-packages/cherrypy/_cpwsgi.py", line 320, in run
    request.run(meth, path, qs, rproto, headers, rfile)
  File "/usr/lib/python2.6/site-packages/cherrypy/_cprequest.py", line 603, in run
    raise cherrypy.TimeoutError()
TimeoutError

I read the salt-api documentation in:
http://salt-api.readthedocs.org/en/latest/ref/netapis/all/saltapi.netapi.rest_cherrypy.html

but i dont see timeout parameter to configure.

Any ideas to solve this problem?
Any information would be appreciated

Regards.

**Comment by whiteinge** [comment]
You may be interested in the [expire_responses](https://docs.saltstack.com/en/latest/ref/netapi/all/salt.netapi.rest_cherrypy.app.html#module-salt.netapi.rest_cherrypy.app) setting. It looks like Rundeck is already including the `timeout` Salt parameter.

Given the nature of HTTP, most people run jobs asynchronously (via the `local_async` or other `*_async` clients) and either poll for job returns or watch the HTTP stream for live returns via the `/events` endpoint. IMO the latter would be a good pattern for Rundeck to adopt. There may be room for some general-purpose salt-api optimizations for long connections but there's a definitely limit there. The `/events` endpoint is specifically tailored for long-running HTTP connections and would be a better fit.

In short, I'm not sure how to help with the immediate problem. The Salt Minions are still happily running the job even after the timeout, and they still return the result of the job to the Salt Master when they're done. But I'm not familiar enough with Rundeck to know if it knows to fetch those results after-the-fact. Does Rundeck have an option to run jobs async?

**Comment by albertmfb** [comment]
updating of rundeck and saltstack the last versions and the expire_responses param has solved the problem.

rundeck-2.6.1-1.12.GA.noarch
salt-api-2015.8.1-1.el6.noarch
salt-master-2015.8.1-1.el6.noarch

Thanks.
